src/tree/splitter.ts

```diff
@@ -0,0 +1,206 @@
+import { Criterion, ImpurityMeasure, SampleData } from './criterion'
+import { shuffle } from 'lodash'
+import { quickSort, deepCopy } from './utils'
+import { int } from '../randUtils'
+
+export class Split {
+  feature: int = 0
+  threshold = 0
+  pos = -1
+  impurity_left: number = Number.POSITIVE_INFINITY
+  impurity_right: number = Number.POSITIVE_INFINITY
+  left_value: int[] = []
+  right_value: int[] = []
+  found_split = false
+}
+
+export class Splitter {
+  kMinSplitDiff_: number
+  feature_data_: number[][]
+  label_data_: int[]
+  criterion_: Criterion
+  start_: int
+  end_: int
+  min_samples_leaf_: int
+  max_features_: int
+  feature_order_: int[]
+  shuffle_features_: boolean
+
+  sample_map_: SampleData[]
+  n_samples_total_: int
+  n_features_: int
+  constructor(
+    feature_data: number[][],
+    label_data: int[],
+    min_samples_leaf: int,
+    impurity_measure: ImpurityMeasure,
+    max_features: int,
+    samples_subset: int[] = []
+  ) {
+    this.feature_data_ = feature_data
+    this.label_data_ = label_data
+    this.n_features_ = feature_data[0].length
+    this.min_samples_leaf_ = min_samples_leaf
+    this.max_features_ = max_features
+    this.shuffle_features_ = max_features < this.n_features_
+    this.sample_map_ = []
+    this.start_ = 0
+    this.end_ = 0
+    this.kMinSplitDiff_ = 1e-8
+    if (samples_subset.length === 0) {
+      this.n_samples_total_ = feature_data.length
+      for (let i = 0; i < this.n_samples_total_; i++) {
+        this.sample_map_.push({ current_feature_value: 0, sample_number: i })
+      }
+    } else {
+      this.n_samples_total_ = samples_subset.length
+      for (let i = 0; i < this.n_samples_total_; i++) {
+        this.sample_map_.push({
+          current_feature_value: 0,
+          sample_number: samples_subset[i]
+        })
+      }
+    }
+    this.criterion_ = new Criterion(impurity_measure, label_data)
+    this.feature_order_ = []
+    for (let i = 0; i < this.n_features_; i++) {
+      this.feature_order_.push(i)
+    }
+    this.resetSampleRange(0, this.n_samples_total_)
+  }
+
+  resetSampleRange(start: int, end: int) {
```

Comment:
I agree with this. I'm basing a lot of this structure of the sklearn implementation here. https://github.com/scikit-learn/scikit-learn/tree/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/tree

and so I'm a little conflicted by diverging too much with their implementation. Let's talk through this more on a call, because I agree with you that merging those functions is better.